### PR TITLE
Clean up extraneous window inset call

### PR DESCRIPTION
**Pull request by GaryQian**
## Description

Fixes bug part of https://github.com/flutter/flutter/issues/65821

`onStart` is called too late, which resulted in one of the final IME position calls to slip through. This PR switches to the `onPrepare` instead and adds some additional safeguarding to prevent overwriting the saved inset. This is consistent with https://github.com/android/user-interface-samples/blob/master/WindowInsetsAnimation/app/src/main/java/com/google/android/samples/insetsanimation/RootViewDeferringInsetsCallback.kt which also uses `onPrepare` and is what this is roughly based on.

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 1]
```diff
@@ -294,6 +298,10 @@ public void onEnd(WindowInsetsAnimation animation) {
         }
       }
     }
+
+    public boolean isAnimating() {
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:302]: Checking the animation state combined with getVisibility() on the IME inset can be used to determine if keyboard is visible without resorting to heuristics, will implement in a separate PR.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 2]
```diff
@@ -199,7 +199,8 @@ public void sendAppPrivateCommand(String action, Bundle data) {
 
     private View view;
     private WindowInsets lastWindowInsets;
-    private boolean started = false;
+    private boolean animating = false;
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:202]: I'd add code comments for what these 2 are for the next maintainer

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 3]
```diff
@@ -212,34 +213,35 @@ public void sendAppPrivateCommand(String action, Bundle data) {
     @Override
     public WindowInsets onApplyWindowInsets(View view, WindowInsets windowInsets) {
       this.view = view;
-      if (started) {
+      if (needsSave) {
+        // Store the view and insets for us in onEnd() below
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:217]: You described the what, but didn't describe the why (saving is needed) :)

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java [hunk 4]
```diff
@@ -677,6 +677,10 @@ public void ime_windowInsetsSync() {
     builder.setInsets(WindowInsets.Type.navigationBars(), Insets.of(10, 10, 10, 40));
     WindowInsets imeInsets1 = builder.build();
```
Comment (xster) [inline shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:678]: add some comments for what 0, 1, 2 are supposed to represent? Different stages of an animation?

**Review by xster — COMMENTED** [review]
LGTM

**Comment by GaryQian** [comment]
Landing, Linux Web Engine passes on rerun: https://ci.chromium.org/p/flutter/builders/try/Linux%20Web%20Engine/9773

**Comment by pcsosinski** [comment]
removing cp label from PR in favor of the issue

**Pull request merged**